rs/recovery/src/steps.rs

```diff
@@ -559,71 +586,71 @@ impl Step for UploadAndRestartStep {
-        // does not exist on the affected version.
-        replace_state.push_str("(sudo systemctl restart setup-permissions || true);");
-        replace_state.push_str("sudo systemctl start ic-replica;");
-        replace_state.push_str("sudo systemctl status ic-replica;");
+        // Decide: remote or local recovery
+        if let UploadMethod::Remote(node_ip) = self.upload_method {
+            // For remote recoveries, we copy the source directory via rsync.
+            // To improve rsync times, we copy the latest checkpoint to the
+            // upload directory.
+
+            let ic_checkpoints_path = format!("{}/{}", IC_DATA_PATH, IC_CHECKPOINTS_PATH);
+            // path of highest checkpoint on upload node
+            let copy_from = format!(
+                "{}/$(ls {} | sort | tail -1)",
+                ic_checkpoints_path, ic_checkpoints_path
+            );
+            // path and name of checkpoint after replay
+            let copy_to = format!("{}/{}/{}", upload_dir, CHECKPOINTS, max_checkpoint);
+            let cp = format!("sudo cp -r {} {}", copy_from, copy_to);
+            let cmd_create_and_copy_checkpoint_dir = format!(
+                "sudo mkdir -p {}/{}; {}; sudo chown -R {} {};",
+                upload_dir, CHECKPOINTS, cp, account, upload_dir
+            );
 
-        ssh_helper.ssh(replace_state)?;
+            let ssh_helper = SshHelper::new(
+                self.logger.clone(),
+                account.to_string(),
+                node_ip,
+                self.require_confirmation,
+                self.key_file.clone(),
+            );
+            info!(
+                self.logger,
+                "Creating remote directory and copying previous checkpoint..."
+            );
+            if let Some(res) = ssh_helper.ssh(cmd_create_and_copy_checkpoint_dir)? {
+                info!(self.logger, "{}", res);
+            }
+            let target = format!("{}@[{}]:{}/", account, node_ip, upload_dir);
+            info!(self.logger, "Uploading state...");
+            rsync(
+                &self.logger,
+                IC_STATE_EXCLUDES.to_vec(),
+                &src,
+                &target,
+                self.require_confirmation,
+                self.key_file.as_ref(),
+            )?;
+            info!(self.logger, "Restarting replica...");
+            ssh_helper.ssh(cmd_replace_state)?;
+        } else {
+            // For local recoveries we simply `mv` state to the upload directory.
+            // No rsync is needed, and thus no checkpoint copying.
+            info!(self.logger, "Moving state locally...");
+            let mut mv_to_target = Command::new("sudo");
+            mv_to_target.arg("mv");
+            mv_to_target.arg(src);
+            mv_to_target.arg(upload_dir);
+            exec_cmd(&mut mv_to_target)?;
+            info!(self.logger, "Restarting replica...");
+            exec_cmd(Command::new("bash").arg("-c").arg(cmd_replace_state))?;
```

Comment:
Do I understand correctly that here we replace the original state with the replayed one? I think we should still preserve the original state until the cleanup step was executed

Reply:
Yeah, I was just mirroring what we do in the remote case. Should we add a prompt asking if we want to persist the existing state, just like the download step?

Reply:
The original state should already be persisted (assuming we made a copy of it during the download state and used that to generate the recovery state). I just think we should only delete/replace the original state when executing the cleanup step. The [ticket](https://dfinity.atlassian.net/browse/CON-1346) should have a rough description of what I mean

Reply:
I made the changes to copy into a backup directory. PTAL

Reply:
Great, thanks! Instead of creating a copy, would it also work to just move the original state into the backup directory?

Reply:
done

Reply:
In this case, would we also have to adapt the "replace state" script? IIUC, now there is no state to "replace" anymore.